<a href="https://colab.research.google.com/github/descaharefa/Introduction-to-Python-Beginner-/blob/main/Study_Case_Introduction_to_Python(Beginner)_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>



# Study Case Introduction to Python

- Nama : Desca winta Harefa


In [ ]:
import numpy as np

# Normal

## 1. Operasi Variabel

> Misal kita memiliki toko sepatu. Modal membeli 1 kodi sepatu adalah Rp 1.000.000. Setiap sepatu dijual seharga Rp 80.000. Buatlah variabel modal, kuantitas, dan harga_jual sesuai informasi di atas. Hitung profit yang didapatkan jika semua sepatu berhasil terjual, dan masukkan ke variabel profit.

Saran langkah-langkah pengerjaan
1. Convert terlebih dahulu kodi dalam jumlah barang (1 kodi = xyz buah sepatu)
2. Gunakan operasi matematika sederhana untuk menghitung profit menggunakan tiga buah variabel yang sudah ditentukan

In [ ]:
# Modal membeli 1 kodi sepatu
modal_per_kodi = 1000000  # dalam Rupiah

# Convert 1 kodi sepatu ke dalam jumlah sepatu
sepatu_per_kodi = 12  # 1 kodi = 12 sepatu

# Harga jual per sepatu
harga_jual_per_sepatu = 80000  # dalam Rupiah

# Menghitung profit jika semua sepatu terjual
modal = modal_per_kodi / sepatu_per_kodi  # Modal per sepatu
profit = harga_jual_per_sepatu - modal # Menghitung profit per sepatu
profit1 = profit * 12  # Menghitung semua sepatu, ubah dalam 1 kodi
profit = round(profit1)

Jadi, dari profit di atas, dapat disimpulkan bahwa profit yang didapatkan jika semua sepatu berhasil terjual ialah -Rp. 40.000. Hasil negatif menunjukkan bahwa sepatu dijual dengan harga lebih rendah daripada modal, yang berarti semua sepatu yang dijual dengan harga Rp 80.000 per sepatu, maka akan mengalami kerugian sebesar Rp. 40.000 (kodi).

## 2. Package Python dan Dictionary

> Kita akan coba mengakses data scientific article dari PLOS corpus
    Menggunakan API request dengan kode berikut :
```
import requests
response = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')
```
**perhatikan tanda kutipnya tidak salah ketika meng-copy ke dalam python

Pertanyaannya :
Convert response.content dari request tersebut dari JSON Format ke dalam Dictionary, dan print-kan key di dalamnya apa saja

Saran langkah-langkah pengerjaan
1. Ikuti code dari soal, kemudian ekstrak value dari variable response menggunakan dot (.) content menjadi sebuah variabel baru (dictionary)
2. Gunakan library json, convert hasil content menggunakan json.loads
3. Print keys dari dictionary tersebut


In [ ]:
import requests
response = requests.get('http://api.plos.org/search?q=title:"Drosophila" AND body:"RNA"&fl=id,abstract')

In [ ]:
import json

# Mengonversi response.content menjadi dictionary
data = json.loads(response.content)

In [ ]:
# Mencetak keys dalam dictionary
print(data.keys())

dict_keys(['response'])


Jadi, dari url link di atas, kami menyimpulkan bahwa memiliki keys dalam dictionary ialah response

# Intermediate

## 3. Operasi dalam List

> Buatlah list dengan 10 nama kota di indonesia, kemudian:
1. Urutkan nama kota di dalam list berdasarkan abjad (a)
2. Ambil 4 kota pertama dari list tersebut (yang telah diurutkan) (b)
3. Ambil 3 kota terakhir dari list tersebut (yang telah diurutkan) (c)

Saran langkah-langkah pengerjaan
1. Tidak ada ketentuan nama kota (misal berdasarkan provinsi X atau pulau Y), dibebaskan untuk memilih. Buat nama-nama kota tersebut dalam sebuah list.
2. Gunakan fungsi sorted(X) untuk mengurutkan secara kecil-ke-besar (sesuai abjad), dimana X adalah contoh variabel list pada langkah (a)
3. Gunakan hasil poin (a), urutkan nama kota di dalam list berdasarkan abjad, untuk mengerjakan poin (b) dan (c).

In [ ]:
nama_kota = ["Jakarta", "Surabaya", "Bandung", "Medan", "Semarang", "Makassar", "Palembang", "Bogor", "Yogyakarta", "Malang"] # Membuat list dengan 10 nama kota di Indonesia

In [ ]:
nama_kota.sort() # Urutkan nama kota dalam list berdasarkan abjad

In [ ]:
empat_kota_pertama = nama_kota[:4] # Mengambil 4 kota pertama dari list yang telah diurutkan

In [ ]:
tiga_kota_terakhir = nama_kota[-3:] # Mengambil 3 kota terakhir dari list yang telah diurutkan

In [ ]:
# Mencetak hasilnya
print("Nama kota yang telah diurutkan:")
print(nama_kota)
print("\nEmpat kota pertama:")
print(empat_kota_pertama)
print("\nTiga kota terakhir:")
print(tiga_kota_terakhir)

Nama kota yang telah diurutkan:
['Bandung', 'Bogor', 'Jakarta', 'Makassar', 'Malang', 'Medan', 'Palembang', 'Semarang', 'Surabaya', 'Yogyakarta']

Empat kota pertama:
['Bandung', 'Bogor', 'Jakarta', 'Makassar']

Tiga kota terakhir:
['Semarang', 'Surabaya', 'Yogyakarta']


Jadi, dari operasi list di atas, dapat disimpulkan bahwa dari 10 list nama - nama kota di Indonesia berikut ini ('Jakarta', 'Surabaya', 'Bandung', 'Medan', 'Semarang', 'Makassar', 'Palembang', 'Bogor', 'Yogyakarta', 'Malang') akan diurutkan terlebih dahulu. Kemudian list nama kota yang telah diurutkan akan menghasilkan seperti berikut ini ('Bandung', 'Bogor', 'Jakarta', 'Makassar', 'Malang', 'Medan', 'Palembang', 'Semarang', 'Surabaya', 'Yogyakarta'), setelah itu akan di ambil 4 kota pertama yang telah diurutkan yaitu ('Bandung', 'Bogor', 'Jakarta', 'Makassar') dan 3 kota terakhir yaitu yang telah diurutkan yaitu ('Semarang', 'Surabaya', 'Yogyakarta').

## 4. Function dalam List

> Team sales memiliki list total transaksi dari 10 merchant pada bulan Juni 2020, yaitu:
```
total_transaksi = [ 19, 35, 23, 25, 10, 15, 33, 45, 21, 30 ]
```
1. Trx minimum
2. Rata-rata
3. Trx maksimal

Saran langkah-langkah pengerjaan
1. Copy variabel total_transaksi yang ada di soal ke notebook kita (atau ketik ulang boleh)
2. Kemudian, hitung menggunakan function operasi dalam list yang telah dipelajari di kelas pada topik list. Print hasilnya setiap poin (a), (b), dan (c)

In [ ]:
# List total transaksi dari10 merchant pada bulan Juni 2020
total_transaksi = [19, 35, 23, 25, 10, 15, 33, 45, 21, 30]

# Hitung nilai transaksi minimum
trx_minimum = min(total_transaksi)
print("Transaksi Minimum:", trx_minimum)

# Hitung rata-rata nilai transaksi
rata_rata = sum(total_transaksi) / len(total_transaksi)
print("Rata-rata Transaksi:", rata_rata)

# Hitung nilai transaksi maksimum
trx_maksimum = max(total_transaksi)
print("Transaksi Maksimum:", trx_maksimum)

Transaksi Minimum: 10
Rata-rata Transaksi: 25.6
Transaksi Maksimum: 45


Jadi, dari hasil operasi list di atas, dapat disimpulkan bahwa nilai minimal, max, rata-rata pada data total transaksi pada bulan Juni 2020 diatas ialah nilai minimal yaitu 10, nilai rata - rata 25,6 dan nilai maximum yaitu 45.

# Advanced

## 5. NumPy Array

> Buatlah array dari 5 bilangan ganjil pertama (dimulai dari 1), Kemudian lakukan perintah berikut :
1. Kali-kan array tersebut dengan 10
2. Bagi-kan array (hasil dari bagian a) dengan 3
3. Kali-kan array (hasil dari bagian b) dengan array baru :
`[-45, 43, 3, 56, 5]`
4. Carilah pada index ke-berapa nilai tertinggi dan terendah dari array (hasil dari bagian c)

Saran langkah-langkah pengerjaan
1. HATI-HATI dalam pengerjaan, karena apabila poin sebelumnya salah, maka selanjutnya akan kebawa salah (misal (b) salah, maka poin (c) sampai (d) akan salah)
2. Untuk array 5 bilangan ganjil pertama boleh manual boleh menggunakan fungsi, dimulai dari 1
3. Ikuti perintah soal, gunakan numpy array, jangan list. Sesuaikan dengan permintaan soal pada setiap butir.



In [ ]:
import numpy as np

In [ ]:
array_awal = np.array([1, 3, 5, 7, 9]) # Buat array dengan 5 bilangan ganjil pertama, dimulai dari 1

In [ ]:
array_kali = array_awal * 10 # Kali-kan array tersebut dengan 10
print(array_kali)

[10 30 50 70 90]


In [ ]:
array_bagi = array_kali / 3 # Bagi-kan array hasil dari langkah sebelumnya dengan 3
print(array_bagi)

[ 3.33333333 10.         16.66666667 23.33333333 30.        ]


In [ ]:
array_baru = np.array([-45, 43, 3, 56, 5]) # Buat array baru: [-45, 43, 3, 56, 5]
array_hasil = array_bagi * array_baru # Kali-kan array hasil dari langkah sebelumnya dengan array baru
print(array_hasil)

[-150.          430.           50.         1306.66666667  150.        ]


In [ ]:
index_tertinggi = np.argmax(array_hasil) # Cari index ke-berapa nilai tertinggi dari array hasil
index_terendah = np.argmin(array_hasil) # Cari index ke-berapa nilai terendah dari array hasil

In [ ]:
print("Nilai tertinggi:", array_hasil[index_tertinggi], "ditemukan pada index ke-", index_tertinggi) # Cetak nilai tertinggi beserta indexnya
print("Nilai terendah:", array_hasil[index_terendah], "ditemukan pada index ke-", index_terendah) # Cetak nilai terendah beserta indexnya

Nilai tertinggi: 1306.6666666666665 ditemukan pada index ke- 3
Nilai terendah: -150.0 ditemukan pada index ke- 0


Jadi, dari hasil operasi numpy di atas, dapat disimpulkan bahwa dari 5 bilangan array secara acak yang dimulai dari angka 1 yakni (1,3,5,7,9) akan dilakukan perintah berikut ini yaitu kali-kan array dengan 10, bagi-kan array dengan 3, kali-kan array dengan yang baru, dan mencari pada index ke-berapa nilai tertinggi dan terendah dari array. Sehingga hasil yang didapatkan dari perintah tersebut akan diuraikan sebagai berikut, pertama kali-kan array dengan 10 menghasilkan nilai (10, 30, 50, 70, 90), kedua bagi-kan array dengan 3 menghasilkan nilai (3.33333333, 10, 16.66666667, 23.33333333, 30), ketiga kali-kan array dengan yang baru menghasilkan nilai (-150, 430, 50, 1306.66666667, 150), dan keempat mencari pada index ke-berapa nilai tertinggi dan terendah dari array menghasilkan nilai tertinggi 1306.66666667 ditemukan pada index ke-3 serta nilai terendah -150 ditemukan pada index ke-0.

# Terimakasih, salam dari ... Medan